### ISSUE TITLE
S5 Slide Show Self-Contained Broken in 2.0.1

### ISSUE THREAD

**Issue by grodri** [jgm/pandoc#4026]
The S5 engine uses four CSS files (and the first one includes three others), plus one JavaScript file. Pandoc generates links to the CSS files that have `media` and `id` attributes:

```
<link rel="stylesheet" href="./s5/default/slides.css" type="text/css" media="projection" id="slideProj" />
<link rel="stylesheet" href="./s5/default/outline.css" type="text/css" media="screen" id="outlineStyle" />
<link rel="stylesheet" href="./s5/default/print.css" type="text/css" media="print" id="slidePrint" />
<link rel="stylesheet" href="./s5/default/opera.css" type="text/css" media="projection" id="operaFix" />
```

In Pandoc 1.9.2 `--self-contained` uses base64 encoding, so the first link becomes
```
<link href="data:text/css;...encoded CSS..." rel="stylesheet" type="text/css" media="projection" id="slideProj" />
```

This works fine. In Pandoc 2.0.1 `--self-contained` includes the CSS code in a style tag:

```
<style type="text/css">...CSS here...</style>
```

but omits the `media` and `id` attributes. This causes problems, because the script in `slides.js` refers to
these elements using `getElementById()`. 

Unfortunately, adding the missing attributes is not enough, because the script also refers to the `href` attribute of `slides.css`, and of course the `style` tag has no such attribute.  Short of rewriting the JavaScript code to avoid this reference, reverting to the 1.9.2 behavior seems the simplest solution.

**Comment by jgm** [comment]
See #3423 for the motivation for the change.

Maybe I'll change the code so that base64 is used if the link tag has a `media` attribute. That should be enough for this case, anyway.

**Comment by grodri** [comment]
Makes perfect sense. I much prefer text for CSS and JS over bulky base64. Adding the id and media attributes to the style tag would solve most of the problem. Looking into tweaking S5 to resolve the href issue.

**Comment by jgm** [comment]
I have changed it to use base64 for this particular case,
so S5 doesn't need to be tweaked.

+++ Germán Rodríguez [Nov 02 17 00:46 ]:

**Comment by grodri** [comment]
Probably the safest tack at this time. But a bit more testing shows that if you add the id and media attributes to the `style` tags everything will work *provided* you change `projection` to `screen`. I am puzzled why `projection` appears to work with a link but not a style tag, but it is deprecated anyway. This means you could go with text instead of base64 as per #3423. 

On a related issue, the S5 script `slides.js` has an IE fix that adds a behavior from file  `ui/default/iepngfix.htc`, which 2.0.1 keeps as a link.  Moreover, the htc file has a link to a blank image at `ui/default/blank.gif`. Microsoft ended support for IE in Jan 2016, so this is really not an issue.

**Comment by jgm** [comment]
+++ Germán Rodríguez [Nov 02 17 16:20 ]:
>   Probably the safest tack at this time. But a bit more testing shows
>   that if you add the id and media attributes to the style tags
>   everything will work provided you change projection to screen. I am

What do you mean, "change projection to screen"?  Can you
tell me the precise change to make?  (It's not a change in
S5 itself, I assume?)

>   puzzled why projection appears to work with a link but not a style tag,
>   but it is deprecated anyway. This means you could go with text instead
>   of base64 as per [1]#3423.
>
>   On a related issue, the S5 script slides.js has an IE fix that adds a
>   behavior from file ui/default/iepngfix.htc, which 2.0.1 keeps as a
>   link. Moreover, the htc file has a link to a blank image at
>   ui/default/blank.gif. Microsoft ended support for IE in Jan 2016, so
>   this is really not an issue.

Again, I don't want to mess with S5 itself.

**Comment by grodri** [comment]
I did not mean a change in S5, and the way things stand, I still think the safest tack is to use ugly base64.
But let me clarify.  The `style` tag substitute for the first link would be
```
<style type="text/css" media="projection" id="slideProj"> ... CSS here ... </style>
```
The `id` is essential because it is used in code to toggle slide and outline views. But this doesn't work. I got better results coding `media="screen"` instead of `media="projection"`, which is deprecated. But just now I ran into an example where this didn't work 100%. Not clear why loading CSS from an external file should be different from including it in the document, but that in conjunction with the media attribute may alter priorities in the cascade. 

On the second issue, I was not suggesting messing with S5 itself.  Just noting the fact that the script `slides.js` includes a link to an external `htc` file.  This should probably be base64 encoded, even if the script itself is in plain text, unless we don't care about IE.

I think I'll run these issues by the author of S5.
